content/blog/nielx/2022-12-16_http_network_services_preview_in_r1_beta_4.md

```diff
@@ -0,0 +1,109 @@
+
+In addition to these features, the implementation of the kit also uses some other modern language features internally, like [initializer lists](https://en.cppreference.com/w/cpp/language/constructor), the `auto` keyword for variable declaration, `constexpr`, and [string literals](https://en.cppreference.com/w/cpp/language/string_literal).
+
+### Exceptions for Error Handling
+
+The new HTTP Network Services Kit uses exceptions for error handling, as an alternative to the traditional way of error handling in the APIs inherited from BeOS. In the current public API, error codes are the way to identify and handle errors: functions and methods that may fail have a `status_t` return type, as well as the use of negative values in `ssize_t` or `off_t` return values to signal errors. For many objects, there is an `InitCheck()` method that should be called in order to check if the object is valid and working correctly. For consistency reasons, all new public APIs and most other code do not use exceptions. There is some precedent for experimentation with exceptions, for example in the [private parts of the package kit](https://git.haiku-os.org/haiku/tree/src/kits/package/hpkg?h=r1beta4).
+
+The preview of the HTTP Network Kit requires the user of the API to do all error handling through exceptions. It has the following characteristics:
+* All error handling is done through exceptions. This means that no method will return `status_t` or overload any other return value to signal an error has occured. It will no longer be necessary to call `InitCheck()` after creating a new object, any issues in the constructor will raise an exception.
+* The API consistently shows which methods and functions can throw exceptions, by the consistent use of the `noexcept` keyword. This is also part of the documentation.
+* The API documentation explicitly states which types of exception can be thrown, as to allow the user of the API to determine which exceptions to handle.
+* The API introduces the [`BPrivate::Network::BError`](https://www.haiku-os.org/docs/api/classBPrivate_1_1Network_1_1BError.html) abstract base class, which defines a standard API for all derived exceptions. Part of this base class is the `origin` property, which holds the location of where the exception occured in order to aid debugging.
+* Next to the exceptions derived from `BError`, users of the API will also find exceptions of the `std::bad_alloc` type for when allocations fail due to the fact that the system is low on memory.
```

Comment:
this seems a bit unpractical to handle. Should we consider using std::nothrow with new and then throwing our own BError derived memory allocation error?

Reply:
In theory we could, though I think BError needs to be redesigned to make sure it does not allocate anything itself. For example, the `origin` parameter is internally stored as a BString, which admittedly is mostly out of convenience for me, so I think making it const char should be possible. Additionally, the DebugMessage() should be rewritten so that it does not allocate. But it is possible.

In my opinion though, the bigger picture is that there are very few applications that need to handle low/no resource scenarios. In most cases you might as well crash the application in the scenarios. It takes a lot of effort to write resilient code that keeps running during those scenarios. We must do that in the servers and other daemons, as well as Tracker and Deskbar, but beyond that...

Reply:
I think that by the time you are running out of resources, even emergency save procedures are perilous and do not have the best odds of succeeding. I think autosave is a much better strategy (and guards against more types of crashes than out of memory). But you're right, another topic for discussion :-)